File: bindings/node/stores/spanner/test/spanner.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { Spanner } from "@google-cloud/spanner";
import { StoreError, missingBytes, presentBytes, upsertNode } from "@trail/prolly-node/remote-store";
import { runStoreConformance } from "@trail/prolly-storetest";

import { SPANNER_DDL, SpannerStore, type SpannerItemClient, type SpannerMutation, type SpannerRootRecord, type SpannerTransaction } from "../src/index.ts";

test("Spanner SDK contract", async (suite) => {
  await suite.test("matches Rust DDL and raw-byte layout while satisfying conformance", async () => {
    assert.deepEqual(SPANNER_DDL, [
      "CREATE TABLE ProllyNodes (\n  Cid BYTES(32) NOT NULL,\n  Node BYTES(MAX) NOT NULL\n) PRIMARY KEY (Cid)",
      "CREATE TABLE ProllyHints (\n  Namespace BYTES(MAX) NOT NULL,\n  HintKey BYTES(MAX) NOT NULL,\n  Value BYTES(MAX) NOT NULL\n) PRIMARY KEY (Namespace, HintKey)",
      "CREATE TABLE ProllyRoots (\n  Name BYTES(MAX) NOT NULL,\n  Manifest BYTES(MAX) NOT NULL\n) PRIMARY KEY (Name)",
    ]);
    const client = new MemoryClient(); const store = SpannerStore.fromClient(client); await runStoreConformance(() => store);
    const cid = specialBytes(32); await store.putNode(cid, bytes("value")); assert.deepEqual(client.node(cid), bytes("value")); assert.deepEqual(client.lastMutation, { kind: "upsertNode", key: cid, value: bytes("value") });
    const descriptor = await store.descriptor(); assert.equal(descriptor.capabilities.nativeBatchReads, false); assert.equal(descriptor.capabilities.atomicBatchWrites, true); assert.equal(descriptor.limits.maxTransactionOperations, undefined);
  });

  await suite.test("serializable CAS has one winner and strict conflicts roll back", async () => {
    const client = new MemoryClient(); const store = SpannerStore.fromClient(client);
    const results = await Promise.all(Array.from({ length: 32 }, (_, index) => store.compareAndSwapRootManifest(bytes("race"), missingBytes(), presentBytes(bytes(`winner-${index}`)))));
    assert.equal(results.filter(({ applied }) => applied).length, 1);
    const conflict = await store.commitTransaction([upsertNode(bytes("rollback-node"), bytes("must-not-write"))], [{ name: bytes("race"), expected: missingBytes() }], [{ kind: "put", name: bytes("rollback-root"), manifest: bytes("must-not-publish") }]);
    assert.equal(conflict.applied, false); assert.equal((await store.getNode(bytes("rollback-node"))).present, false); assert.equal((await store.getRootManifest(bytes("rollback-root"))).present, false);
  });

  await suite.test("maps retryable errors without leaking provider details", async () => {
    const secret = "spanner-service-account-secret";
    const client = new MemoryClient(); client.failure = Object.assign(new Error(secret), { code: 14 }); const store = SpannerStore.fromClient(client);
    await assert.rejects(store.putNode(bytes("cid"), bytes("value")), (error: unknown) => { assert.ok(error instanceof StoreError); assert.equal(error.code, "unavailable"); assert.equal(error.retryable, true); assert.doesNotMatch(error.message, new RegExp(secret)); return true; });
  });

  await suite.test("propagates cancellation and preserves client ownership", async () => {
    let closed = false;
    const client: SpannerItemClient & { close(): void } = {
      close(): void { closed = true; }, async getNode() { return missingBytes(); }, async getHint() { return missingBytes(); }, async getRoot() { return missingBytes(); }, async listNodeCids() { return []; }, async listRoots() { return []; },
      async apply(_mutations, signal) { await new Promise((_resolve, reject) => signal?.addEventListener("abort", () => reject(Object.assign(new Error("aborted"), { code: 1 })), { once: true })); },
      async readWrite<T>(_callback: (transaction: SpannerTransaction) => Promise<T>): Promise<T> { throw new Error("unused"); },
    };
    const store = SpannerStore.fromClient(client); const controller = new AbortController(); const write = store.putNode(bytes("cid"), bytes("value"), controller.signal); controller.abort("test");
    await assert.rejects(write, (error: unknown) => { assert.ok(error instanceof StoreError); assert.equal(error.code, "cancelled"); return true; }); await store.close(); assert.equal(closed, false);
  });
});

test("Spanner emulator conformance", { skip: process.env.SPANNER_EMULATOR_HOST === undefined }, async () => {
  const projectId = "prolly-test"; const instanceId = "prolly-test"; const databaseId = `prolly_node_${Date.now()}`; const spanner = new Spanner({ projectId }); const instance = spanner.instance(instanceId);
  const [exists] = await instance.exists(); if (!exists) { const [, operation] = await spanner.createInstance(instanceId, { config: "emulator-config", nodes: 1 }); await operation.promise(); }
  const [database, operation] = await instance.createDatabase(databaseId, { schema: [...SPANNER_DDL] }); await operation.promise(); const store = new SpannerStore(database);
  try { await runStoreConformance(() => store); } finally { await store.close(); await database.close(); await database.delete(); spanner.close(); }
});

interface State { nodes: Map<string, Uint8Array>; hints: Map<string, Uint8Array>; roots: Map<string, Uint8Array>; }
class MemoryClient implements SpannerItemClient {
  #state: State = emptyState(); #tail: Promise<void> = Promise.resolve(); failure?: Error; lastMutation?: SpannerMutation;
  async getNode(key: Uint8Array): Promise<ReturnType<typeof missingBytes>> { return optional(this.#state.nodes.get(id(key))); }
  async getHint(namespace: Uint8Array, key: Uint8Array): Promise<ReturnType<typeof missingBytes>> { return optional(this.#state.hints.get(`${id(namespace)}:${id(key)}`)); }
  async getRoot(name: Uint8Array): Promise<ReturnType<typeof missingBytes>> { return optional(this.#state.roots.get(id(name))); }
  async listNodeCids(): Promise<Uint8Array[]> { return [...this.#state.nodes.keys()].map(fromId); }
  async listRoots(): Promise<SpannerRootRecord[]> { return [...this.#state.roots].map(([name, manifest]) => ({ name: fromId(name), manifest: own(manifest) })); }
  async apply(mutations: readonly SpannerMutation[]): Promise<void> { if (this.failure !== undefined) { const failure = this.failure; this.failure = undefined; throw failure; } const next = cloneState(this.#state); for (const mutation of mutations) apply(next, mutation); this.#state = next; this.lastMutation = mutations.at(-1); }
  async readWrite<T>(callback: (transaction: SpannerTransaction) => Promise<T>): Promise<T> {
    let release!: () => void; const previous = this.#tail; this.#tail = new Promise<void>((resolve) => { release = resolve; }); await previous;
    try { const next = cloneState(this.#state); const buffered: SpannerMutation[] = []; const result = await callback({ getRoot: async (name) => optional(next.roots.get(id(name))), buffer: (mutations) => buffered.push(...mutations.map(cloneMutation)) }); for (const mutation of buffered) apply(next, mutation); this.#state = next; return result; } finally { release(); }
  }
  node(key: Uint8Array): Uint8Array | undefined { const value = this.#state.nodes.get(id(key)); return value === undefined ? undefined : own(value); }
}

function emptyState(): State { return { nodes: new Map(), hints: new Map(), roots: new Map() }; }
function cloneState(state: State): State { return { nodes: new Map([...state.nodes].map(([key, value]) => [key, own(value)])), hints: new Map([...state.hints].map(([key, value]) => [key, own(value)])), roots: new Map([...state.roots].map(([key, value]) => [key, own(value)])) }; }
function apply(state: State, mutation: SpannerMutation): void { switch (mutation.kind) { case "upsertNode": state.nodes.set(id(mutation.key), own(mutation.value)); break; case "deleteNode": state.nodes.delete(id(mutation.key)); break; case "upsertHint": state.hints.set(`${id(mutation.namespace)}:${id(mutation.key)}`, own(mutation.value)); break; case "upsertRoot": state.roots.set(id(mutation.key), own(mutation.value)); break; case "deleteRoot": state.roots.delete(id(mutation.key)); break; } }
function cloneMutation(value: SpannerMutation): SpannerMutation { switch (value.kind) { case "upsertNode": return { kind: value.kind, key: own(value.key), value: own(value.value) }; case "deleteNode": return { kind: value.kind, key: own(value.key) }; case "upsertHint": return { kind: value.kind, namespace: own(value.namespace), key: own(value.key), value: own(value.value) }; case "upsertRoot": return { kind: value.kind, key: own(value.key), value: own(value.value) }; case "deleteRoot": return { kind: value.kind, key: own(value.key) }; } }
function optional(value: Uint8Array | undefined): ReturnType<typeof missingBytes> { return value === undefined ? missingBytes() : presentBytes(value); }
function id(value: Uint8Array): string { return Buffer.from(value).toString("base64"); }
function fromId(value: string): Uint8Array { return own(Buffer.from(value, "base64")); }
function own(value: Uint8Array): Uint8Array { return Uint8Array.from(value); }
function bytes(value: string): Uint8Array { return own(Buffer.from(value)); }
function specialBytes(length: number): Uint8Array { const pattern = [0, 0x7f, 0x80, 0xff]; return Uint8Array.from(Array.from({ length }, (_, index) => pattern[index % pattern.length]!)); }

File: bindings/go/stores/spanner/spanner_test.go
package spanner

import (
	"bytes"
	"context"
	"encoding/binary"
	"errors"
	"fmt"
	"os"
	"sort"
	"strings"
	"sync"
	"testing"
	"time"

	prolly "build.crab/prolly-go"
	"build.crab/prolly-go/storetest"
	gspanner "cloud.google.com/go/spanner"
	database "cloud.google.com/go/spanner/admin/database/apiv1"
	databasepb "cloud.google.com/go/spanner/admin/database/apiv1/databasepb"
	instance "cloud.google.com/go/spanner/admin/instance/apiv1"
	instancepb "cloud.google.com/go/spanner/admin/instance/apiv1/instancepb"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func TestDDLMatchesRust(t *testing.T) {
	ddl := strings.Join(DDLStatements, ";\n")
	for _, fragment := range []string{
		"CREATE TABLE ProllyNodes", "Cid BYTES(32) NOT NULL", "Node BYTES(MAX) NOT NULL", "PRIMARY KEY (Cid)",
		"CREATE TABLE ProllyHints", "Namespace BYTES(MAX) NOT NULL", "HintKey BYTES(MAX) NOT NULL", "Value BYTES(MAX) NOT NULL", "PRIMARY KEY (Namespace, HintKey)",
		"CREATE TABLE ProllyRoots", "Name BYTES(MAX) NOT NULL", "Manifest BYTES(MAX) NOT NULL", "PRIMARY KEY (Name)",
	} {
		if !strings.Contains(ddl, fragment) {
			t.Fatalf("DDL missing %q", fragment)
		}
	}
}

func TestNewDefaults(t *testing.T) {
	store := New(nil, Options{})
	_, err := store.Descriptor(context.Background())
	if err == nil {
		t.Fatal("nil SDK client should be rejected")
	}
	var _ prolly.RemoteStore = store
}

func TestSpannerSDKContractConformance(t *testing.T) {
	store := NewWithClient(newMemoryClient(), Options{})
	storetest.RunWithStore(t, prolly.RemoteStore(store))
}

func TestSpannerEmulatorConformance(t *testing.T) {
	if os.Getenv("SPANNER_EMULATOR_HOST") == "" {
		t.Skip("SPANNER_EMULATOR_HOST is not set")
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()
	projectID, instanceID := "prolly-test", "prolly-test"
	instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer instanceAdmin.Close()
	instanceName := fmt.Sprintf("projects/%s/instances/%s", projectID, instanceID)
	operation, err := instanceAdmin.CreateInstance(ctx, &instancepb.CreateInstanceRequest{
		Parent: "projects/" + projectID, InstanceId: instanceID,
		Instance: &instancepb.Instance{Config: "projects/" + projectID + "/instanceConfigs/emulator-config", DisplayName: instanceID, NodeCount: 1},
	})
	if err == nil {
		if _, err := operation.Wait(ctx); err != nil {
			t.Fatal(err)
		}
	} else if status.Code(err) != codes.AlreadyExists {
		t.Fatal(err)
	}
	databaseAdmin, err := database.NewDatabaseAdminClient(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer databaseAdmin.Close()
	databaseID := fmt.Sprintf("prolly_go_%d", time.Now().UnixNano())
	databaseName := instanceName + "/databases/" + databaseID
	databaseOperation, err := databaseAdmin.CreateDatabase(ctx, &databasepb.CreateDatabaseRequest{
		Parent: instanceName, CreateStatement: "CREATE DATABASE `" + databaseID + "`", ExtraStatements: append([]string(nil), DDLStatements...),
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := databaseOperation.Wait(ctx); err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		_ = databaseAdmin.DropDatabase(context.Background(), &databasepb.DropDatabaseRequest{Database: databaseName})
	})
	client, err := gspanner.NewClient(ctx, databaseName)
	if err != nil {
		t.Fatal(err)
	}
	t.Cleanup(client.Close)
	storetest.RunWithStore(t, prolly.RemoteStore(New(client, Options{})))
}

func TestSpannerErrorClassification(t *testing.T) {
	for _, test := range []struct {
		code      codes.Code
		retryable bool
	}{
		{code: codes.Aborted, retryable: true},
		{code: codes.Unavailable, retryable: true},
		{code: codes.ResourceExhausted, retryable: true},
		{code: codes.FailedPrecondition, retryable: false},
	} {
		err := spannerError("test", status.Error(test.code, "failure"))
		var storeErr *prolly.StoreError
		if !errors.As(err, &storeErr) || storeErr.Retryable != test.retryable || storeErr.ProviderCode != test.code.String() {
			t.Fatalf("code %s: error = %#v", test.code, err)
		}
	}
}

func TestConcurrentRootCASAppliesExactlyOnce(t *testing.T) {
	store := NewWithClient(newMemoryClient(), Options{})
	results := make(chan prolly.RootCASResult, 2)
	errorsChannel := make(chan error, 2)
	var wait sync.WaitGroup
	for _, value := range [][]byte{[]byte("one"), []byte("two")} {
		wait.Add(1)
		go func(value []byte) {
			defer wait.Done()
			result, err := store.CompareAndSwapRootManifest(context.Background(), []byte("root"), prolly.MissingBytes(), prolly.PresentBytes(value))
			results <- result
			errorsChannel <- err
		}(value)
	}
	wait.Wait()
	close(results)
	close(errorsChannel)
	for err := range errorsChannel {
		if err != nil {
			t.Fatal(err)
		}
	}
	applied := 0
	for result := range results {
		if result.Applied {
			applied++
		}
	}
	if applied != 1 {
		t.Fatalf("applied = %d", applied)
	}
}

type memoryState struct {
	nodes map[string][]byte
	hints map[string][]byte
	roots map[string][]byte
}

type memoryClient struct {
	mu    sync.Mutex
	state memoryState
}

func newMemoryClient() *memoryClient {
	return &memoryClient{state: emptyState()}
}

func emptyState() memoryState {
	return memoryState{nodes: make(map[string][]byte), hints: make(map[string][]byte), roots: make(map[string][]byte)}
}

func (c *memoryClient) GetNode(ctx context.Context, key []byte) (prolly.OptionalBytes, error) {
	if err := ctx.Err(); err != nil {
		return prolly.OptionalBytes{}, err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	return optional(c.state.nodes[string(key)]), nil
}

func (c *memoryClient) GetHint(ctx context.Context, namespace, key []byte) (prolly.OptionalBytes, error) {
	if err := ctx.Err(); err != nil {
		return prolly.OptionalBytes{}, err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	return optional(c.state.hints[hintMapKey(namespace, key)]), nil
}

func (c *memoryClient) GetRoot(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	if err := ctx.Err(); err != nil {
		return prolly.OptionalBytes{}, err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	return optional(c.state.roots[string(name)]), nil
}

func (c *memoryClient) ListNodeCIDs(ctx context.Context) ([][]byte, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([][]byte, 0, len(c.state.nodes))
	for key := range c.state.nodes {
		result = append(result, []byte(key))
	}
	sort.Slice(result, func(i, j int) bool { return bytes.Compare(result[i], result[j]) < 0 })
	return result, nil
}

func (c *memoryClient) ListRoots(ctx context.Context) ([]RootRecord, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([]RootRecord, 0, len(c.state.roots))
	for name, manifest := range c.state.roots {
		result = append(result, RootRecord{Name: []byte(name), Manifest: clone(manifest)})
	}
	sort.Slice(result, func(i, j int) bool { return bytes.Compare(result[i].Name, result[j].Name) < 0 })
	return result, nil
}

func (c *memoryClient) Apply(ctx context.Context, mutations []Mutation) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	state := cloneState(c.state)
	applyMutations(&state, mutations)
	c.state = state
	return nil
}

func (c *memoryClient) ReadWrite(ctx context.Context, function func(context.Context, Transaction) error) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	state := cloneState(c.state)
	if err := function(ctx, &memoryTransaction{state: &state}); err != nil {
		return err
	}
	c.state = state
	return nil
}

type memoryTransaction struct{ state *memoryState }

func (t *memoryTransaction) GetRoot(ctx context.Context, name []byte) (prolly.OptionalBytes, error) {
	if err := ctx.Err(); err != nil {
		return prolly.OptionalBytes{}, err
	}
	return optional(t.state.roots[string(name)]), nil
}

func (t *memoryTransaction) Buffer(mutations []Mutation) error {
	applyMutations(t.state, mutations)
	return nil
}

func applyMutations(state *memoryState, mutations []Mutation) {
	for _, mutation := range mutations {
		switch mutation.Kind {
		case MutationUpsertNode:
			state.nodes[string(mutation.Key)] = clone(mutation.Value)
		case MutationDeleteNode:
			delete(state.nodes, string(mutation.Key))
		case MutationUpsertHint:
			state.hints[hintMapKey(mutation.Namespace, mutation.Key)] = clone(mutation.Value)
		case MutationUpsertRoot:
			state.roots[string(mutation.Key)] = clone(mutation.Value)
		case MutationDeleteRoot:
			delete(state.roots, string(mutation.Key))
		}
	}
}

func cloneState(state memoryState) memoryState {
	result := emptyState()
	for key, value := range state.nodes {
		result.nodes[key] = clone(value)
	}
	for key, value := range state.hints {
		result.hints[key] = clone(value)
	}
	for key, value := range state.roots {
		result.roots[key] = clone(value)
	}
	return result
}

func optional(value []byte) prolly.OptionalBytes {
	if value == nil {
		return prolly.MissingBytes()
	}
	return prolly.PresentBytes(value)
}

func hintMapKey(namespace, key []byte) string {
	var length [8]byte
	binary.BigEndian.PutUint64(length[:], uint64(len(namespace)))
	return string(length[:]) + string(namespace) + string(key)
}
